### ISSUE TITLE
Citation Keys mixed up

### ISSUE THREAD

**Issue by cjpoor** [retorquere/zotero-better-bibtex#1919]
**Zotero version:**
5.0.97-beta.43+c5d89f6d0

**BBT version:**
5.4.29

**Support log ID:**
7JLDXGI5-refs-apse

**Expected behavior:**
I would expect this export to look like this:
```

@article{singh_indigenous_2013,
  title = {Indigenous Knowledge and Pedagogy for {{Indigenous}} Children},
  author = {Singh, Navin Kumar and Reyhner, Jon},
  date = {2013},
  journaltitle = {Honoring our children: Culturally appropriate approaches for teaching Indigenous students},
  pages = {37--52},
  file = {/media/cjpoor/Seagate Expansion Drive/zotero_data/storage/JH2GJD8S/Indigenous Knowledge and Pedagogy.pdf}
}
``` 
But instead, I am regularly finding that Citation Keys are for the wrong item. 
**Actual behavior:** For example:
```

@article{abensour_against_2007,
  title = {Indigenous Knowledge and Pedagogy for {{Indigenous}} Children},
  author = {Singh, Navin Kumar and Reyhner, Jon},
  date = {2013},
  journaltitle = {Honoring our children: Culturally appropriate approaches for teaching Indigenous students},
  pages = {37--52},
  file = {/media/cjpoor/Seagate Expansion Drive/zotero_data/storage/JH2GJD8S/Indigenous Knowledge and Pedagogy.pdf}
}
```

I noticed that there is another entry with author Singh and year 2013 and this also has an incorrect Citation key.

**Support log ID:**
P59KDUUK-refs-apse

**Actual behavior:**
```

@article{gillespie_media_2005,
  title = {Research Assessments and Rankings: {{Accounting}} for Accountability in "{{Higher Education Ltd}}"},
  author = {Singh, Geeta},
  date = {2013},
  journaltitle = {International Education Journal: Comparative Perspectives},
  volume = {9},
  number = {1},
  file = {/media/cjpoor/Seagate Expansion Drive/zotero_data/storage/GZ3GUX8F/Singh_2013_Research assessments and rankings.pdf}
}
```

**Expected behavior:**
I would expect this export to look like this:
```

@article{singh_research_2013a,
  title = {Research Assessments and Rankings: {{Accounting}} for Accountability in "{{Higher Education Ltd}}"},
  author = {Singh, Geeta},
  date = {2013},
  journaltitle = {International Education Journal: Comparative Perspectives},
  volume = {9},
  number = {1},
  file = {/media/cjpoor/Seagate Expansion Drive/zotero_data/storage/GZ3GUX8F/Singh_2013_Research assessments and rankings.pdf}
}
```

Verified

*Please verify **all** of the below before submitting*:

1. you have picked the right category for the issue in the previous screen.
2. You are on the latest release of Zotero
3. in the Zotero addons screen you can see that you have the latest release of BBT (https://github.com/retorquere/zotero-better-bibtex/releases/latest)
4. you are posting a single bug or feature request.
5. the issue has a subject that succinctly describes the problem or question.
6. you are available for follow-up questions and testing.
7. you have included an support log ID here generated by reproducing the problem, selecting the problematic reference(s), right-clicking, and submitting an BBT support log from that popup menu, **not** from the `Help` menu.
8. you have included the *actual output* you got from exporting the items you sent in the debug log, not simplified fictional output or a description of the output, and a sample (*not* a description) of how you want it to look instead, again based on the actual items you selected in the previous step.

Picking the right issue category is really important. Each category (`Export`, `General error`, `Import`, `Key generation`, `Question`) has different instructions for gathering the data necessary required to resolve the issue you are experiencing

The support log is important; it gives me your current BBT settings and a copy of the problematic reference as a test case so I can best replicate your problem. Without it, I'm effectively blind.

**Comment by retorquere** [comment]
Can you send a debug log for both, and check the "include cache" checkbox before submitting?

**Comment by cjpoor** [comment]
Singh article with abensour key:

Support log reporting tool threw this:

[Exception... "Illegal value"  nsresult: "0x80070057 (NS_ERROR_ILLEGAL_VALUE)"  location: "JS frame :: chrome://zotero-better-bibtex/content/better-bibtex.js :: send :: line 83524"  data: no] (7ENPXXHR-refs-apse, references: true)

and then froze on "Please wait while the support log is submitted"

On the second try it worked with 
***Support Log ID*** 
XD9CH499-refs-apse


Singh article with gillespie key

***Support Log ID***
9Z5T44J4-refs-apse

**Comment by retorquere** [comment]
I've never seen the `NS_ERROR_ILLEGAL_VALUE` error, but unfortunately that is an error that I cannot handle -- these `NS_` errors simply aborts the running code, even the error handler.

I phrased myself poorly in the previous message -- I meant, can you select both , right-click one of them, and then send a log? That way they both en up in the same log.

**Comment by cjpoor** [comment]
OK
***Support Log ID***
TP3KIHZI-refs-apse

**Comment by retorquere** [comment]
Can you install [this translatoe](https://gist.github.com/ce4ce50b8ddf38ea6492e8cb503208dd), export those two items, and attach the result here?

**Comment by cjpoor** [comment]
Result attached
[Exported_Items.csv](https://github.com/retorquere/zotero-better-bibtex/files/7131872/Exported_Items.csv)

**Comment by cjpoor** [comment]
These are the two Singh items with incorrect citation keys - attached.
[Exported_Items1.csv](https://github.com/retorquere/zotero-better-bibtex/files/7131907/Exported_Items1.csv)

**Comment by cjpoor** [comment]
There are more if you want me to send them.

**Comment by retorquere** [comment]
Can you send a new error report and check the "include cache" option?

**Comment by cjpoor** [comment]
For which items?

**Comment by retorquere** [comment]
Good point -- it doesn't matter in this case, you can send it from the Help menu.

**Comment by cjpoor** [comment]
From the Help menu there was no option to include cache.
ID: 860512489

**Comment by cjpoor** [comment]
From the Singh items it hung up submitting report log. On second attempt gave error: TypeError: wizard.advance is not a function (MJ83ZR3C-refs-apse, references: true)

**Comment by retorquere** [comment]
> From the Help menu there was no option to include cache.
> ID: 860512489

That's a Zotero log. There should also be an option "Send BBT Support log" in the help menu.

**Comment by cjpoor** [comment]
This was actually sent from the Help menu - Just the two original files
ID  412702701

**Comment by retorquere** [comment]
BBT support log IDs always contain `-euc` or `-apse`. IDs that do not have these are sent to Zotero, I have no access to those logs. The exact menu item under the Help menu is `Send Better BibTeX support log...`

**Comment by cjpoor** [comment]
Now I keep getting:
TypeError: wizard.advance is not a function (4TXG4IGR-refs-apse, references: true)
When trying to send Better BibTeX support log

**Comment by retorquere** [comment]
Can you save the log locally (there should be a button)?

**Comment by cjpoor** [comment]
But eventually the log seems to be sent.
latest ID: ADIUAKXU-refs-apse

**Comment by cjpoor** [comment]
Yes. I saved one yesterday, and this just now: ID 96X8B55N-refs-apse

**Comment by retorquere** [comment]
This won't fix the citekeys but may give me more info on why error reports are faulting. I still don't know what causes the key mixup.

**Comment by cjpoor** [comment]
OK. Support log ID: KM3DKN5Y-refs-apse

**Comment by cjpoor** [comment]
In the translators folder there is an old file with this strange "s" with a subscript "." in the suffix:
```
Better BibTeX.js
```

**Comment by cjpoor** [comment]
I don't know if it's relevant, but I recently moved my Zotero data folder to an external hard drive.

**Comment by retorquere** [comment]
`KM3DKN5Y-refs-apse` has

```
Unix error 21 during operation read on file /media/cjpoor/Seagate Expansion Drive/zotero_data/storage/TJUPU3VK/re-thinking-history-120.pdfd
```

but no other errors. But the only thing that build 1513 would allow you to do differently is if the log upload fails with the `wizard.advance` problem, you can generate a *new* log, save that locally, and attach the resulting zip here. The new log should have captured the error that occured when the `wizard.advance` problem happened.



> strange "s" with a subscript "." in the suffix:

No idea where this could come from. It's not something BBT would, or even could do. I hand the translators to Zotero, and Zotero appends the `.js` extension. 

> I don't know if it's relevant, but I recently moved my Zotero data folder to an external hard drive.

It shouldn't be relevant, as long as Zotero wasn't running when you moved it.

**Comment by cjpoor** [comment]
I just made a correction to the title of one of the Singh items and the key immediately corrected itself.

**Comment by cjpoor** [comment]
I had previously tried adding a letter to the author name and then deleting it. That had no effect.

**Comment by retorquere** [comment]
The author name should also have triggered a refresh, there must be an error in the log for that. But if you just want to get this fixed, selecting the entries, right-click, and BBT>refresh keys will do it. I honestly have no idea how your situation can come about, I'm looking through the previous logs to see if there is some kind of mismatch between item IDs and keys, but they all check out so far.

**Comment by cjpoor** [comment]
Right. Refresh key fixes them. I notice that you have had a long struggle to make these keys compatible with Zotero. 
The first issue I saw was in 2014 - raised by a user called Singh! :-0

**Comment by cjpoor** [comment]
A BibTeX library that I saved in Feb 2020 has these incorrect citation keys (and all those saved later) but one saved in Dec 2019 does not appear to have any.

**Comment by retorquere** [comment]
If a refresh fixes it then the problem is that the keys were assigned to the wrong item, but I have *no* idea how that could happen, and if we can't reproduce it, I'm afraid there's nothing I can do right now to diagnose it.

**Comment by cjpoor** [comment]
I remember that I had to restore from a duplicity backup around the time this appeared. From what Dan says about backing up it seems possible that the database I restored from was corrupted by my backup process. I may have been backing up from two computers to the one cloud storage (although not syncing).

I was also using Syncthing around then, trying to keep Zotero data in sync on two computers. 

I also may have overwritten the Zotero data folder contents with the backup contents rather than deleting the original folder and replacing with the backup.

I now sync everything on Zotero server :-)
